core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java

```diff
@@ -58,7 +58,6 @@ protected HandlerResult doAuthentication(final Credential credential) throws Gen
         }
 
         userPass.setUsername(transformedUsername);
-        userPass.setPassword(this.passwordEncoder.encode(userPass.getPassword()));
```

Comment:
How about to add this to 

```
        userPass.setUsername(transformedUsername);
             // encode only if BCRYPT is not used. BCRYPT option should be
             // handled individually in handlers using matches
        if(!(this.passwordEncoder instanceof BCryptPasswordEncoder)){
            userPass.setPassword(this.passwordEncoder.encode(userPass.getPassword()));    
        }        
        return authenticateUsernamePasswordInternal(userPass);
```

 This way all will work as it is until now and we just need to add routine like matches which i did so any authentication handler can do logic like

 ```
if(! (this.matches() || <original way of check>)){
   throw new FailedLoginException("Password does not match value on record.");
}
```

this way it is minimum impact on existing and we just need to ensure that there is test with BCRYPT for all possible handlers and fix the issue